blob/fileblob/fileblob.go

```diff
@@ -57,35 +59,18 @@ func OpenBucket(dir string) (*blob.Bucket, error) {
 	return blob.NewBucket(&bucket{dir}), nil
 }
 
-// resolvePath converts a key into a relative filesystem path. It guarantees
-// that there will only be one valid key for a given path and that the resulting
-// path will not reach outside the directory.
-func resolvePath(key string) (string, error) {
-	for _, c := range key {
-		if !('A' <= c && c <= 'Z' || 'a' <= c && c <= 'z' || '0' <= c && c <= '9' || c == '/' || c == '.' || c == ' ' || c == '_' || c == '-') {
-			return "", fmt.Errorf("contains invalid character %q", c)
-		}
-	}
-	if cleaned := slashpath.Clean(key); key != cleaned {
-		return "", fmt.Errorf("not a clean slash-separated path")
-	}
-	if slashpath.IsAbs(key) {
-		return "", fmt.Errorf("starts with a slash")
-	}
-	if key == "." {
-		return "", fmt.Errorf("invalid path \".\"")
-	}
-	if strings.HasPrefix(key, "../") {
-		return "", fmt.Errorf("starts with \"../\"")
-	}
-	return filepath.FromSlash(key), nil
+func escape(s string) string {
+	s = url.PathEscape(s)
```

Comment:
My concern still here is that we need a specific property that URL escaping usually does not provide, which is that given the same input string, the escaping always produces the exact same string. If `PathEscape` truly provides this, let's document it here, but if not, we should have an explicitly stable naming algorithm.

If the latter, percent-encoding is still fine, but we should make the whitelisted characters explicit.

Reply:
Done.

Per meatspace discussion, during unescaping I now also return an error if `escape(unescape(s)) != s`, so that `blob` is internally consistent.

Also added tests.